service/gcs/core/gcs/gcs.go

```diff
@@ -468,35 +471,29 @@ func (c *gcsCore) RunExternalProcess(params prot.ProcessParameters, stdioSet *co
 		}
 		utils.LogMsgf("external process %d exited with exit status %d", cmd.Process().Pid(), cmd.ExitState().ExitCode())
 
-		// Close stdin so that the copying goroutine is safely unblocked; this is necessary
-		// because the host expects stdin to be closed before it will report process
-		// exit back to the client, and the client expects the process notification before
-		// it will close its side of stdin (which io.Copy is waiting on in the copying goroutine).
-		if stdioSet.In != nil {
-			if err := stdioSet.In.CloseRead(); err != nil {
-				logrus.Errorf("failed call to CloseRead for external process stdin: %v: %s", ociProcess.Args, err)
+		if master != nil {
```

Comment:
The reason before why stdioSet.In was always closed on process exit was because when the process exits, the stdin socket isn't automatically closed. Any copy operations which are hung on a read from that socket never return, because as far as they know more data could still come over the socket. As a result, the relays on the host don't get properly cleaned up. Does this problem not occur when a file is duped and passed in directly? Just want to make sure this is a scenario you tested.